expression/util.go

```diff
@@ -66,11 +67,32 @@ func ColumnSubstitute(expr Expression, schema *Schema, newExprs []Expression) Ex
 func datumsToConstants(datums []types.Datum) []Expression {
 	constants := make([]Expression, 0, len(datums))
 	for _, d := range datums {
-		constants = append(constants, &Constant{Value: d})
+		constants = append(constants, &Constant{Value: d, RetType: types.NewFieldType(kindToMysqlType[d.Kind()])})
 	}
 	return constants
 }
 
+var kindToMysqlType = map[byte]byte{
```

Comment:
@coocood 
1. I'm not sure which are not correct, here I map `types.KindXXX` to `mysql.TypeXXX` following the similar rules used in `mysql.TypeXXX.ToClass()`.

2. kindToMysqlType is used by `datumsToConstants` which is used by many test files,
the kinds we used *often* are KindInt64/KindUint64/KindFloat32/KindFloat64/KindString/KindMysqlTime,
but it can not be sure the other types would not be used in the future, so I think it may be better to keep all the mappings.